### PR TITLE
/midi static typing finished

### PR DESCRIPTION
**Pull request by szgorski**
Full static typing for /midi provided:

- reassignment of values then dynamically changing types
- full type specification of in/out function arguments
- full referenced arguments sterilization (no longer modified inside fns - copy.deepcopy applied where necessary
- test updated accordingly

That was **not** done:

- modification of bach.py (3 typing errors still cause midi-ci to fail in last run i.e. 24)
- full type specification of internal fn variables (only if necessary)

### CODE REVIEW

**Comment by szgorski** [comment]
and sorry for 16 commits :<

File: midi/midi/decode.py [hunk 1]
```diff
@@ -18,25 +20,26 @@ class EventNote:
 
     def __init__(
             self,
-            velocity: int | float,
+            velocity: Union[int, float],
```
Comment (piotrowskv) [inline midi/midi/decode.py:23]: Why _Union_ instead of | ?

**Review by piotrowskv — APPROVED** [review]

File: midi/midi/decode.py (hunk 1 above)
Comment (shilangyu) [inline midi/midi/decode.py:23]: Why this change?

File: midi/midi/decode.py [hunk 2]
```diff
@@ -18,25 +20,26 @@ class EventNote:
...
                self.tone == other.tone and \
                self.octave == other.octave
 
-    def __repr__(self):
+    def __repr__(self) -> str:
         return f'EventNote({self.velocity}, {self.height})'
 
     def normalise(self,
-                  max_velocity: int | float):
+                  max_velocity: Union[int, float]) -> None:
```
Comment (shilangyu) [inline midi/midi/decode.py:42]: etc

File: midi/midi/decode.py [hunk 3]
```diff
@@ -350,17 +362,18 @@ def export_output(folder: str,
...
     translates all Messages to a single list, then removes vacuous Messages
 
     :param tracks:
     :return:
     """
     note_grid = [0] * 128                            # list with numbers of active notes, one for each height
-    raw_messages = list[tuple[int, Message]]()       # all messages with their starting times
-    filtered_messages = list[tuple[int, Message]]()  # as raw_messages, but without repetitions
+    raw_messages = list[Tuple[int, Message]]()       # all messages with their starting times
+    filtered_messages = list[Tuple[int, Message]]()  # as raw_messages, but without repetitions
```
Comment (shilangyu) [inline midi/midi/decode.py:374]: Why this change?

File: midi/midi/music_21.py [hunk 4]
```diff
@@ -1,40 +1,45 @@
...
     
     :param array: 
     :return: 
     """
-    events = [[] for _ in range(len(array) + 1)]
-    tuples = [None] * len(array)
+    array = copy.deepcopy(array)
+    events: list[list[list[int]]] = [[] for _ in range(len(array))]
+    tuples = dict[int, Tuple[int, list[int]]]()
+    # tuples: list[Tuple[int, list[int]]] = [] * len(array)
```
Comment (shilangyu) [inline midi/midi/music_21.py]: debug?

File: midi/midi/decode.py (hunk 1 above)
Comment (szgorski) [reply midi/midi/decode.py:23]: it turns out a component (don't remember which one) fails at complex structures to differentiate logical "|" from mathematical "|" (CI threw a few errors)

Comment (szgorski) [reply midi/midi/decode.py:23]: it turns out a component (don't remember which one) fails at complex structures to differentiate logical "|" from mathematical "|" (CI threw a few errors)

File: midi/midi/decode.py (hunk 2 above)
Comment (szgorski) [reply midi/midi/decode.py:42]: it turns out a component (don't remember which one) fails at complex structures to differentiate logical "|" from mathematical "|" (CI threw a few errors)

File: midi/midi/decode.py (hunk 3 above)
Comment (szgorski) [reply midi/midi/decode.py:374]: it turns out a component (don't remember which one) fails at complex tuples to correctly count the number of arguments in a tuple (CI threw a few errors)

File: midi/midi/music_21.py (hunk 4 above)
Comment (szgorski) [reply midi/midi/music_21.py]: corrected in dc1a838446b74d0fcf07c89894ed4ef042469233

**Review by shilangyu — APPROVED** [review]
